service/src/main/java/bio/terra/workspace/service/job/JobBuilder.java

```diff
@@ -116,29 +126,80 @@ public JobBuilder addParameter(String keyName, @Nullable Object val) {
   /**
    * Submit a job to stairway and return the jobID immediately.
    *
+   * @param changeType Specify the type of {@link ActivityLogChangedType} to be logged.
+   * @param resultClass type for the response of the flight. When null, nothing is returned from the
+   *     flight.
    * @return jobID of submitted flight
    */
-  public String submit() {
+  public <T> String submit(
+      @Nullable ActivityLogChangedType changeType, @Nullable Class<T> resultClass) {
     populateInputParams();
-    return jobService.submit(flightClass, jobParameterMap, jobId);
+    if (changeType == null) {
+      return jobService.submit(flightClass, jobParameterMap, jobId);
+    }
+    return jobService.submitWithCallback(
+        flightClass,
+        resultClass,
+        jobParameterMap,
+        jobId,
+        new FutureCallback<>() {
+          @Override
+          public void onSuccess(JobResultOrException<T> result) {
+            if (changeType.equals(ActivityLogChangedType.DELETE)) {
+              activityLogDao.setChangedDate(workspaceId, changeType);
+            } else {
+              if (result.getException() == null) {
+                activityLogDao.setChangedDate(workspaceId, changeType);
+              }
+            }
+          }
+
+          @Override
+          public void onFailure(Throwable t) {}
+        });
   }
 
   /**
    * Submit a job to stairway, wait until it's complete, and return the job result.
    *
    * @param resultClass Class of the job's result
+   * @param doAccessCheck whether to check user request's access to the job.
+   * @param activityLogChangedType type of change activity to be logged when the job finishes. When
+   *     null, nothing is logged.
    * @return Result of the finished job.
    */
   @Traced
-  public <T> T submitAndWait(Class<T> resultClass, boolean doAccessCheck) {
+  public <T> T submitAndWait(
+      Class<T> resultClass,
+      boolean doAccessCheck,
+      @Nullable ActivityLogChangedType activityLogChangedType) {
     populateInputParams();
-    return jobService.submitAndWait(
-        flightClass, jobParameterMap, resultClass, jobId, doAccessCheck);
+    try {
+      T result =
+          jobService.submitAndWait(flightClass, jobParameterMap, resultClass, jobId, doAccessCheck);
+      activityLogDao.setChangedDate(workspaceId, activityLogChangedType);
+      return result;
+    } catch (Exception e) {
+      if (activityLogChangedType != null
+          && activityLogChangedType.equals(ActivityLogChangedType.DELETE)) {
+        // DELETE job cannot be undone. so even when there is failure, we need to set a changed date
```

Comment:
If deletion flight failed, then deletion *maybe* happened, right?

In the DB, I'd like to distinguish between "Deletion definitely happened" and "Deletion maybe happened". Can you log DELETE is job succeeded, and DELETE_MAYBE if job failed?

I don't think we should log DELETE if there's a chance deletion didn't happen.

Reply:
It would be nice if a failed delete job could indicate if resource was deleted or not. (Do a get on the resource.) DELETE_MAYBE would be easier...